README.txt
Dubtel One API - PHP REST Library 0.0.2
============================================

	Thank you for downloading the the PHP REST library for
	Dubtel One API. 

	Description
	===========
	The Dubtel One REST API allows your application to place and 
        receive calls, send and receive fax, send sms as well as 
        query call logs.

        During a call process utilities in the Dubtel One API and 
	API Library will allow your application to perform tasks like:
     
            -- Play Pre-Recorded Audio
            -- Record Audio
            -- Say (Digits/Text)
            -- Receive Input (Digits)
            -- Send Call to Conference Room
            -- Forward Call to any number
            -- Send Fax to Email
            -- Forward Fax to any number

        The Dubtel One Open API also provides IM Presence service, and 
        information on this can be located on the Documentation Site,
	See http://developer.dubtel.com/?q=documentation.

Installation
============

To install this library in your application, just place 
the file dubtel-one.php on your web server and include 
it in the your php file. 


Configuration
=============
To configure your application you will need to specify 
your Dubtel One APPID (Application ID) and the PASSPHRASE
assigned to you when you generated your application key. 

	Sample Codes
	============
	sample-code-outbound.php
	sample-code-outbound-to-vm.php
        sample-code-outbound-to-saydigits.php
        sample-code-outbound-to-playaudio.php
        sample-code-outbound-sms.php

	PHP REST Library
	================
	dubtel-one.php is the main PHP Library, it contains
	the dubtelOneClient Class for handling all 
	Dubtel One API processes.
			

Documentation
=============

   Please refer to the Dubtel One API Documentation for instructions
   and information on getting started, See http://developer.dubtel.com/?q=documentation.


    
How to report bugs
==================

	Please submit all bugs to the Dubtel Support Forum, our support team
	works around the clock to fix all bug issues. If you need immediate
	assitance you can email the Dubtel support team at: support@dubtel.com.

	The Dubtel Support page will allow you to submit new topics and browse
	already submitted issues.

	See: 
	Dubtel Support
	http://developer.dubtel.com/?q=developer

	Dubtel Support Forum
	http://developer.dubtel.com/?q=forum/1


LICENSE
=======
This PHP Library is distributed under the MIT License


RELEASE-NOTES.txt
09-21-2010 : JO : Renamed README and LICENSE files to README.txt and LICENSE.txt so windows users can easily open files after downloads.
09-26-2010 : JO : Changed the DUBTEL Support link in the README.txt from http://developer.dubtel.com/?q=support-and-downloads to http://developer.dubtel.com/?q=developer
09-26-2010 : JO : Changed section name Documenation to Documentation in README.txt
09-26-2010 : JO : Upgraded version number to 0.0.2

dubtel-one.php
<?php

###################################################################################################
#	===========================================================================================
#	Copyright (c) 2010 Dubtel Inc.
#
#	Permission is hereby granted, free of charge, to any person obtaining a copy
#	of this software and associated documentation files (the "Software"), to deal
#	in the Software without restriction, including without limitation the rights
#	to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
#	copies of the Software, and to permit persons to whom the Software is
#	furnished to do so, subject to the following conditions:
#
#	The above copyright notice and this permission notice shall be included in
#	all copies or substantial portions of the Software.
#
#	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#	IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#	FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
#	AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#	LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
#	OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
#	THE SOFTWARE.
#	
#	===========================================================================================
#
# dubtelOneClient API Classes - Dubtel 2010
#
#	The Dubtel One Client Class allows your application to place and 
#       receive calls, send and receive fax, send sms as well as 
#       query call logs.
#
#       During a call process utilities in the Dubtel One API and 
#	API Library will allow your application to perform tasks like:
#     
#            -- Play Pre-Recorded Audio
#            -- Record Audio
#            -- Say (Digits/Text)
#            -- Receive Input (Digits)
#            -- Send Call to Conference Room
#            -- Forward Call to any number
#            -- Send Fax to Email
#            -- Forward Fax to any number
#
#       The Dubtel One Open API also provides IM Presence service, and 
#       information on this can be located on the Documentation Site,
#	See http://www.dubtel.com/?q=documentation.
#
# Public classes
#	dubtelOneClient(APPID, APPKEY, GATEWAY) 
#	-- main Dubtel One API Class
#
#	dubtelOneResponse(URL, RESULT, STATUS, FEATURE) 
#	-- holds all the REST response data from Dubtel One API
#
# Public methods
#     	dubtelOneClient->Send(feature, params) Sends a Request to Service for initiating feature
#	dubtelOneClient->addException(command, params) Creates a Dynamic Extension
#
###################################################################################################

    /* 
     *	Before doing anything lets make sure curl is enable
     */
    if(!extension_loaded("curl"))
        throw(new Exception(
            "Curl extension is required for dubtelOneClient to work, please resolve this."));

    /* 
     * dubtelOneClient throws dubtelOneException on error 
     * Use this to catch dubtelOneClient specific Exceptions in your application
     */
    class dubtelOneException extends Exception {}


    class dubtelOneClient {

	protected $_APPID; 	// Application ID from Dubtel
	protected $_APPKEY;	// PASSPHRASE from Dubtel
	protected $_GATEWAY;    // Dubtel One REST Gateway URL
	protected $_FEATURE;	// Dubtel Feature (Call, Fax, SMS)
	protected $_SSID;	// Dubtel One SSID (Secure Session ID)
 
        public function __construct($applID, $applKey,
            $applgateway = "http://api.dubtel.com") {
            
            $this->_APPID = $applID;
            $this->_APPKEY = $applKey;
            $this->_GATEWAY = $applgateway;
        }

        /*
	 * addExtension
	 *    Creates a dynamic extension in Dubtel One Service for already
	 *	initiated Feature
	 *
	 *    param(s): command, params[n]
	 */
	public function addExtension($ext_command, $params = array())
	{
	    $myparams = $params;
	    $myparams['command'] = $ext_command;
	    return($this->Send($this->_FEATURE,$myparams));	
	}

        /*
         * Send
         *   Sends a Request to the Dubtel One Service to Initiate Feature
         *   $params : an id/value associative array,
	 *           : containing (callerID, Called and GateWay to Call From, etc)
         */
        public function Send($feature, $params = array()) 
	{

            $_entries = "&";
	    $_auth = "&appid={$this->_APPID}&appkey={$this->_APPKEY}";
	    $this->_FEATURE = $feature;

            foreach($params AS $id=>$value)
                $_entries .= "$id=".urlencode($value)."&";
            $_entries = substr($_entries, 0, -1);
            
            // Initialize a new Curl Object
	    $_url = $this->_GATEWAY."/?q=".$feature."{$_auth}{$_entries}";      
            $_ch = curl_init($_url);

            // Authenticate applicatioin request
            curl_setopt($_ch, CURLOPT_USERPWD, $_auth);
	    curl_setopt($_ch, CURLOPT_SSL_VERIFYPEER, false);
	    curl_setopt($_ch, CURLOPT_HTTPAUTH, CURLAUTH_DIGEST);

            curl_setopt($_ch, CURLOPT_RETURNTRANSFER, 1);
            curl_setopt($_ch, CURLOPT_POST, 1);
            curl_setopt($_ch, CURLOPT_POSTFIELDS, $_entries);

            $xmlDoc = new DOMDocument();
	    $xmlDoc->load($_url);

	    $_res = $xmlDoc->saveXML(); 
	    $_httpCode = 200;
	                
            return new dubtelOneResponse($_url, $_res, $_httpCode, $feature);
        }
	
    }  


  /* 
     * dubtelOneResponse holds all the REST response data 
     * Before using the reponse, check isError to see if an exception 
     * occurred with the data sent to Dubtel
     * _responseXml will contain a SimpleXml object with the response xml from Dubtel
     * _responseText contains the raw string response
     * _url and QueryString are from the request
     * _httpStatus is the response code of the request
     * _feature is the Dubtel One API Feature Used (Call, Fax, SMS)
     */
    class dubtelOneResponse {
        
        public $_responseText;
        public $_responseXml;
        public $_httpStatus;
        public $_url;
        public $_queryString;
        public $_isError;
        public $_errorMessage;
	public $_feature;
        
        public function __construct($url, $text, $status, $feature) {
            preg_match('/([^?]+)\??(.*)/', $url, $matches);
            $this->_url = $matches[1];
            $this->_queryString = $matches[2];
            $this->_responseText = $text;
            $this->_httpStatus = $status;
            $this->_feature = $feature;
            if($this->_httpStatus != 204)
                $this->_responseXml = @simplexml_load_string($text);
            
            if($this->_isError = ($status >= 400))
                $this->_errorMessage =
                    (string)$this->_responseXml->dubtelOneResponse->Exception;
            
        }
        
    }    

?>


sample-code-sms.php
<?php
    // Sample Code to Send SMS with the Dubtel One REST API

    // Include the PHP dubtelOne Library
    require "dubtel-one.php";
    
    // Set our Registration Info
    $applID  = "1010101010101010101010";  // Your APPID from Dubtel
    $applKEY = "ATTATTATTATTATTATTATTATT";  // Your PASSPHRASE from Dubtel
    
    // Set CallerID: You can use any CallerID, Leave blank to use your Dubtel One Number
    $callerID = '6145590061';
    
    // Create a new Dubtel One Rest Client Object
    $client = new dubtelOneClient($applID, $applKEY);

    /* 
     * Send an SMS to "614-559-0062"
     */
    $params = array(
    	"Caller" => $callerID, 	      	  // SMS Caller ID (CallerID)
    	"Called" => "6145590062",	  // The Phone number you are sending SMS to
					  // (SMS Recepient)
	"Message" => "Hello World",
    	"GateWay" => "http://api.dubtel.com"
    );
    
    // Send SMS to "6145590062"
    $result = $client->Send("sms", $params);

    // Check result for issues (Success/Error)
    if($result->_isError)
    {
    	echo "Error Sending SMS: {$result->_errorMessage}\n";
	// print_r($result);
    }
    else
    {
    	echo "SMS Sent Successfully: {$result->_responseXml->sms->ssid}\n";
    } 

?>


sample-code-outbound.php
<?php
    // Sample Code to Place OutBound Calls with the Dubtel One REST API

    // Include the PHP dubtelOne Library
    require "dubtel-one.php";
    
    // Set our Registration Info
    $applID  = "1010101010101010101010";  // Your APPID from Dubtel
    $applKEY = "ATTATTATTATTATTATTATTATT";  // Your PASSPHRASE from Dubtel
    
    // Set CallerID: You can use any CallerID, Leave blank to use your Dubtel One Number
    $callerID = '6145590061';
    
    // Create a new Dubtel One Rest Client Object
    $client = new dubtelOneClient($applID, $applKEY);

    /* 
     * Place a new OutBound Call to "614-309-7502"
     */
    $params = array(
    	"Caller" => $callerID, 	      	  // Outgoing Caller ID (CallerID)
    	"Called" => "6145590062",	  // The Phone number you are calling (Calling Number)
    	"GateWay" => "http://api.dubtel.com"
    );
    
    // Place Call to "614-559-0062"
    $result = $client->Send("call", $params);

    // Check result for issues (Success/Error)
    if($result->_isError)
    {
    	echo "Error Placing Phone Call: {$result->_errorMessage}\n";
	print_r($result);
    }
    else
    {
    	print_r($result); // echo "Call Started Successfully: {$result->dubtelOneResponse->call->ssid}\n";
    } 

?>


sample-code-outbound-to-vm.php
<?php
    // Sample Code to Place OutBound Calls with the Dubtel One REST API
    // This code will send called party straight to Voicemail

    // Include the PHP dubtelOne Library
    require "dubtel-one.php";
    
    // Set our Registration Info
    $applID  = "1010101010101010101010";  // Your APPID from Dubtel
    $applKEY = "ATTATTATTATTATTATTATTATT";  // Your PASSPHRASE from Dubtel
    
    // Set CallerID: You can use any CallerID, Leave blank to use your Dubtel One Number
    $callerID = '6145590061';
    
    // Create a new Dubtel One Rest Client Object
    $client = new dubtelOneClient($applID, $applKEY);

    /* 
     * Place a new OutBound Call to "614-559-0062"
     */
    $params = array(
    	"Caller" => $callerID, 	      	  // Outgoing Caller ID (CallerID)
    	"Called" => "6145590062",	  // The Phone number you are calling (Calling Number)
    	"GateWay" => "http://api.dubtel.com"
    );
    
    // Place Call to "614-559-0062"
    $result = $client->Send("call", $params);

    // Check result for issues (Success/Error)
    if($result->_isError)
    {
    	echo "Error Placing Phone Call: {$result->_errorMessage}\n";
    }
    else
    {
    	echo "Call Started Successfully: {$result->_responseXml->call->ssid}\n";
	// print_r($result);
    } 

?>


sample-code-outbound-to-playaudio.php
<?php
    // Sample Code to Place OutBound Calls with the Dubtel One REST API	
    // This Sample will Playback Audio file to Party

    // Include the PHP dubtelOne Library
    require "dubtel-one.php";
    
    // Set our Registration Info
    $applID  = "7234589657746256";  // Your APPID from Dubtel
    $applKEY = "2b47fb821da8ed84b223d678cf22ab03";  // Your PASSPHRASE from Dubtel
    
    // Set CallerID: You can use any CallerID, Leave blank to use your Dubtel One Number
    $callerID = '6145590061';
    
    // Create a new Dubtel One Rest Client Object
    $client = new dubtelOneClient($applID, $applKEY);

    /* 
     * Place a new OutBound Call to "614-559-0062"
     */
    $params = array(
    	"Caller" => $callerID, 	      	  // Outgoing Caller ID (CallerID)
    	"Called" => "6143097502",	  // The Phone number you are calling (Calling Number)
    	"GateWay" => "http://api.dubtel.com"
    );
    
    // Place Call to "614-559-0062"
    $result = $client->Send("call", $params);

    // Check result for issues (Success/Error)
    if($result->_isError)
    {
    	echo "Error Placing Phone Call: {$result->_errorMessage}\n";
    }
    else
    {
	$new_ssid = $result->_responseXml->call->ssid;
    	echo "Call Started Successfully: {$new_ssid}\n";
	// print_r($result); // Uncomment this for debugging
    } 

    /* 
     * Add an Extention to Play an Audio File on OutBound Call
     */
    $params = array(
    	"SSID" => $new_ssid, 	      	  // New SSID From Feature Initiation
    	"audiofilename" => "welcome"	  // Audio File to Play
    );

    // Add Extension to Play AudioFile
    // addExtension(Command, parameters)
    $result = $client->addExtension("Playback", $params);



    // Check result for issues (Success/Error)
    if($result->_isError)
    {
    	echo "Error Generating Extension: {$result->_errorMessage}\n";
    }
    else
    {
    	echo "Extension Generated Successfully: {$result->_responseXml->call->Playback->ssid}\n";
	// print_r($result); // Uncomment this for debugging
    } 

?>


sample-code-outbound-to-saydigits.php
<?php
    // Sample Code to Place OutBound Calls with the Dubtel One REST API
    // This Sample will Say Suggested Digits to Party

    // Include the PHP dubtelOne Library
    require "dubtel-one.php";
    
    // Set our Registration Info
    $applID  = "1010101010101010101010";  // Your APPID from Dubtel
    $applKEY = "ATTATTATTATTATTATTATTATT";  // Your PASSPHRASE from Dubtel
    
    // Set CallerID: You can use any CallerID, Leave blank to use your Dubtel One Number
    $callerID = '6145590061';
    
    // Create a new Dubtel One Rest Client Object
    $client = new dubtelOneClient($applID, $applKEY);

    /* 
     * Place a new OutBound Call to "614-559-0062"
     */
    $params = array(
    	"Caller" => $callerID, 	      	  // Outgoing Caller ID (CallerID)
    	"Called" => "6145590062",	  // The Phone number you are calling (Calling Number)
    	"GateWay" => "http://api.dubtel.com"
    );
    
    // Place Call to "614-559-0062"
    $result = $client->Send("call", $params);


    // Check result for issues (Success/Error)
    if($result->_isError)
    {
    	echo "Error Placing Phone Call: {$result->_errorMessage}\n";
    }
    else
    {
    	echo "Call Started Successfully: {$result->_responseXml->call->ssid}\n";
	// print_r($result); // Uncomment this for debugging
    }


    /*
     * Add an Extention to Play an Audio File on OutBound Call
     */
    $new_ssid = $result->_responseXml->call->ssid;
    $params = array(
    	"SSID" => $new_ssid, 	      	  // New SSID From Feature Initiation
    	"digits" => "12345"	  // Digits to say on call
    );

    // Add Extension to Play AudioFile
    // addExtension(Command, parameters)
    $result = $client->addExtension("SayDigits", $params);



    // Check result for issues (Success/Error)
    if($result->_isError)
    {
    	echo "Error Generating Extension: {$result->_errorMessage}\n";
    }
    else
    {
    	echo "Extension Generated Successfully: {$result->_responseXml->call->SayDigits->ssid}\n";
	// print_r($result); // Uncomment this for debugging
    } 

?>
